File: backend/app/schemas/intelligence.py
"""Schemas for advanced market intelligence APIs."""

from __future__ import annotations

from datetime import date, datetime
from typing import Dict, List, Literal

from pydantic import BaseModel, Field


SignalDirection = Literal["STRONG_BUY", "BUY", "NEUTRAL", "SELL", "STRONG_SELL"]


class TFTSignal(BaseModel):
    p10: float
    p50: float
    p90: float
    raw_signal: float
    horizon_days: int


class HMMGarchSignal(BaseModel):
    regime_signal: str
    vol_forecast_1d: float
    vol_forecast_21d: float


class GNNSignal(BaseModel):
    spillover_risk: float
    embedding_norm: float
    top_correlated_assets: List[str]


class LSTMAttnSignal(BaseModel):
    raw_signal: float
    attention_peaks: List[Dict[str, float]]


class XGBoostSignal(BaseModel):
    raw_signal: float
    top_features: List[Dict[str, float | str]]


class EnsembleSignal(BaseModel):
    signal: SignalDirection
    confidence: float = Field(ge=0.0, le=1.0)
    direction: SignalDirection
    kelly_fraction: float = Field(ge=0.0, le=1.0)


class RegimeDetails(BaseModel):
    state: Literal["BULL", "BEAR", "SIDEWAYS", "CRISIS"]
    probs: Dict[str, float]
    transition_probs: Dict[str, float]


class SignalResponse(BaseModel):
    symbol: str
    timestamp: datetime
    ensemble: EnsembleSignal
    models: Dict[str, TFTSignal | HMMGarchSignal | GNNSignal | LSTMAttnSignal | XGBoostSignal]
    model_weights: Dict[str, float]
    regime: RegimeDetails


class SignalHistoryPoint(BaseModel):
    timestamp: datetime
    signal: float
    actual_return: float
    model: str


class RegimeCurrentResponse(BaseModel):
    state: Literal["BULL", "BEAR", "SIDEWAYS", "CRISIS"]
    probs: Dict[str, float]
    transition_matrix: List[List[float]]
    cond_vol_1d: float
    cond_vol_5d: float
    cond_vol_21d: float
    days_in_state: int
    last_transition_date: date


class RegimeHistoryPoint(BaseModel):
    time: date
    state: Literal["BULL", "BEAR", "SIDEWAYS", "CRISIS"]
    probs: Dict[str, float]
    cond_vol: float


class RegimeStatisticsItem(BaseModel):
    state: Literal["BULL", "BEAR", "SIDEWAYS", "CRISIS"]
    avg_duration: float
    avg_return: float
    avg_vol: float
    freq: float


class SHAPContribution(BaseModel):
    name: str
    shap_val: float
    feature_val: float
    pct_rank: float


class ExplainShapResponse(BaseModel):
    model: Literal["xgboost"]
    feature_contributions: List[SHAPContribution]
    base_value: float
    output_value: float
    waterfall_ready: bool


class AttentionTopTimestep(BaseModel):
    date: date
    weight: float


class ExplainAttentionResponse(BaseModel):
    model: Literal["tft", "lstm_attn"]
    weights: List[List[float]]
    mean_weights: List[float]
    top_timesteps: List[AttentionTopTimestep]


class CounterfactualFeatureChange(BaseModel):
    name: str
    original: float
    counterfactual: float


class CounterfactualResponse(BaseModel):
    cf_id: str
    changed_features: List[CounterfactualFeatureChange]
    resulting_signal: SignalDirection
    proximity_score: float


class CounterfactualRequest(BaseModel):
    target_direction: Literal["BUY", "SELL"]
    num_cfs: int = Field(ge=1, le=10, default=5)


class ModelAccuracyItem(BaseModel):
    model: Literal["tft", "hmm_garch", "gnn", "lstm_attn", "xgboost", "ensemble"]
    precision: float
    recall: float
    directional_accuracy: float
    p50_rmse: float
    winkler_coverage_score: float


class DriftItem(BaseModel):
    model: Literal["tft", "hmm_garch", "gnn", "lstm_attn", "xgboost", "ensemble"]
    adwin_p_value: float
    drift_detected: bool
    residual_distribution: List[float]
    ks_stat: float


class EnsembleWeightPoint(BaseModel):
    date: date
    tft: float
    hmm_garch: float
    gnn: float
    lstm_attn: float
    xgboost: float


class EconomicCalendarEvent(BaseModel):
    date: date
    category: Literal["FOMC", "EARNINGS", "EXPIRY", "MACRO"]
    title: str
    impact: Literal["LOW", "MEDIUM", "HIGH"]


class CorrelationNode(BaseModel):
    ticker: str
    sector: str
    x: float
    y: float
    size: float


class CorrelationEdge(BaseModel):
    source: str
    target: str
    correlation: float


class CorrelationGraphResponse(BaseModel):
    window_days: int
    central_asset: str
    nodes: List[CorrelationNode]
    edges: List[CorrelationEdge]
    top_correlates: List[Dict[str, float | str]]


class FactorExposureItem(BaseModel):
    factor: str
    beta: float


class FactorExposureResponse(BaseModel):
    symbol: str
    window_days: int
    exposures: List[FactorExposureItem]
    metrics: Dict[str, float]

File: apps/web/src/types/intelligence.ts
export type SignalDirection =
  | "STRONG_BUY"
  | "BUY"
  | "NEUTRAL"
  | "SELL"
  | "STRONG_SELL";

export interface EnsembleSignal {
  signal: SignalDirection;
  confidence: number;
  direction: SignalDirection;
  kelly_fraction: number;
}

export interface RegimeDetails {
  state: "BULL" | "BEAR" | "SIDEWAYS" | "CRISIS";
  probs: Record<string, number>;
  transition_probs: Record<string, number>;
}

export interface TFTSignal {
  p10: number;
  p50: number;
  p90: number;
  raw_signal: number;
  horizon_days: number;
}

export interface HMMGarchSignal {
  regime_signal: string;
  vol_forecast_1d: number;
  vol_forecast_21d: number;
}

export interface GNNSignal {
  spillover_risk: number;
  embedding_norm: number;
  top_correlated_assets: string[];
}

export interface LSTMAttnSignal {
  raw_signal: number;
  attention_peaks: Array<Record<string, number>>;
}

export interface XGBoostSignal {
  raw_signal: number;
  top_features: Array<Record<string, string | number>>;
  xgb_confidence?: number;
  xgb_direction?: string;
  train_samples?: number;
}

export interface TechnicalModelSignal {
  score?: number;
  rsi?: number;
  macd_histogram?: number;
  bb_position?: number;
  adx?: number;
  supertrend_direction?: number;
  above_vwap?: boolean;
  indicators_computed?: number;
  atr?: number;
  ema9?: number;
  ema21?: number;
  ema50?: number;
  ema200?: number;
}

export interface PatternModelSignal {
  pattern_score?: number;
  patterns_detected?: string[];
  bullish_count?: number;
  bearish_count?: number;
}

export interface MomentumModelSignal {
  momentum_score?: number;
  ret_1d?: number;
  ret_5d?: number;
  ret_21d?: number;
  jt_momentum?: number;
  vol_21d?: number;
  yang_zhang_vol?: number;
  dist_52w_high?: number;
  dist_52w_low?: number;
}

export interface RegimeModelSignal {
  regime?: string;
  bull_prob?: number;
  bear_prob?: number;
  regime_confidence?: number;
  hmm_used?: boolean;
}

export interface SignalResponse {
  symbol: string;
  timestamp: string;
  ensemble: EnsembleSignal;
  models: {
    tft: TFTSignal;
    hmm_garch: HMMGarchSignal;
    gnn: GNNSignal;
    lstm_attn: LSTMAttnSignal;
    xgboost: XGBoostSignal;
    technical?: TechnicalModelSignal;
    pattern?: PatternModelSignal;
    momentum?: MomentumModelSignal;
    regime?: RegimeModelSignal;
  };
  model_weights: Record<string, number>;
  regime: RegimeDetails;
  target_price_5d?: number;
  stop_loss?: number;
  take_profit?: number;
  prob_buy?: number;
  prob_sell?: number;
  max_loss_pct?: number;
  is_computed?: boolean;
  message?: string;
  model_status?: "trained" | "heuristic" | "fallback";
}

export interface ForecastPoint {
  target_date: string;
  horizon_days: number;
  predicted_price: number;
  predicted_direction: string;
  confidence: number;
  prediction_low: number;
  prediction_high: number;
  change_pct?: number;
}

export interface ModelForecast {
  model_name: string;
  forecasts: ForecastPoint[];
}

export interface ForecastResponse {
  symbol: string;
  current_price: number;
  generated_at: string;
  model_results: ModelForecast[];
  forecast?: ForecastPoint[];
}
